cluster-operator/src/test/java/io/strimzi/operator/cluster/operator/resource/KafkaMetadataStateManagerTest.java

```diff
@@ -0,0 +1,179 @@
+
+import java.util.Map;
+
+import static io.strimzi.api.kafka.model.kafka.KafkaMetadataState.KRaft;
+import static io.strimzi.api.kafka.model.kafka.KafkaMetadataState.KRaftDualWriting;
+import static io.strimzi.api.kafka.model.kafka.KafkaMetadataState.KRaftMigration;
+import static io.strimzi.api.kafka.model.kafka.KafkaMetadataState.KRaftPostMigration;
+import static io.strimzi.api.kafka.model.kafka.KafkaMetadataState.PreKRaft;
+import static io.strimzi.api.kafka.model.kafka.KafkaMetadataState.ZooKeeper;
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.junit.jupiter.api.Assertions.assertThrows;
+
+/**
+ * Tests the state transitions which happens in the KafkaMetadataStateManager class.
+ */
+public class KafkaMetadataStateManagerTest {
+
+    private static final String CLUSTER_NAMESPACE = "my-namespace";
+
+    private static final String CLUSTER_NAME = "kafka-test-cluster";
+
+    private static final Reconciliation RECONCILIATION = new Reconciliation("test-trigger", Kafka.RESOURCE_KIND, CLUSTER_NAMESPACE, CLUSTER_NAME);
+
+    private static final Kafka KAFKA = new KafkaBuilder()
+            .withNewMetadata()
+                .withName(CLUSTER_NAME)
+                .withAnnotations(Map.of(Annotations.ANNO_STRIMZI_IO_NODE_POOLS, "enabled"))
+                .withNamespace(CLUSTER_NAMESPACE)
+            .endMetadata()
+            .withNewSpec()
+                .withNewKafka()
+                    .withListeners(new GenericKafkaListenerBuilder()
+                        .withName("plain")
+                        .withPort(9092)
+                        .withType(KafkaListenerType.INTERNAL)
+                        .withTls(false)
+                    .build())
+                .endKafka()
+            .endSpec()
+            .build();
+
+    /**
+     *  Computes the next state to which the previous state has transitioned
+     *
+     * @param kafka The Kafka instance.
+     * @param status Status of the Kafka custom resource where warnings about any issues with metadata state will be added
+     * @param isKRaftFeatureGateEnabled if the UseKRaft feature gate is enabled on the operator
+     *
+     * @return the state to which the previous state has transitioned
+     */
+    private KafkaMetadataState computeNextTransition(Kafka kafka, KafkaStatus status, boolean isKRaftFeatureGateEnabled) {
+        KafkaMetadataStateManager kafkaMetadataStateManager = new KafkaMetadataStateManager(RECONCILIATION, kafka, isKRaftFeatureGateEnabled);
+        return kafkaMetadataStateManager.computeNextMetadataState(status);
+    }
+
+    @Test
+    public void testMoveFromZookeeperToKRaftMigrationWhenMigrationAnnotationEnabled() {
+
+        Kafka kafka = new KafkaBuilder(KAFKA)
+                .editMetadata()
+                    .addToAnnotations(Annotations.ANNO_STRIMZI_IO_KRAFT, "migration")
+                .endMetadata()
+                .withNewStatus()
+                    .withKafkaMetadataState("ZooKeeper")
```

Comment:
I think it would be better to use enum here like `KafkaMetadataState.ZooKeeper.name()`.
It's possible their values are going to change during the review so if we have hard coded string here, these tests start to fail.
Please do the same in all the other places.